packages/api/test/repository-test.js

```diff
@@ -59,7 +64,99 @@ describe('Repository', () => {
     });
   });
 
-  describe('performActionIfAllowed', async () => {
-    it('performs requested test delete when user has write permission to repo');
+  describe('ticket handlers', async () => {
+    let state;
+    const ticket = {
+      data: 'fake data'
+    };
+    beforeEach(() => {
+      state = v4();
+      ticket.state = state;
+    });
+    describe('storeTicket', async () => {
+      it('stores a ticket which has a state', async () => {
+        // Store ticket
+        await repo.storeTicket(ticket);
+
+        // Retrieve ticket, check it
+        const result = await repo.getDoc(`tickets/${state}`);
+        assert.deepStrictEqual(result.ticket, ticket);
+
+        // Delete ticket
+        await repo.deleteDoc(`tickets/${state}`);
+      });
+    });
+
+    describe('getTicket', async () => {
+      it('retrieves a ticket by its state', async () => {
+        await repo.createDoc(`tickets/${state}`, {
+          ticket: ticket
+        });
+
+        const result = await repo.getTicket(state);
+
+        assert.deepStrictEqual(result, ticket);
+
+        await repo.deleteDoc(`tickets/${state}`);
+      });
+    });
+  });
+
+  describe('allowedToPerformTicket', async () => {
+    it('returns true when user has write permission to repo and attempts to delete a test', async () => {
+      const result = await repo.allowedToPerformTicket('delete-test', 'write');
+      assert.deepStrictEqual(result, true);
+    });
+    it('returns true when user has admin permission to repo and attempts to delete a test', async () => {
+      const result = await repo.allowedToPerformTicket('delete-test', 'admin');
+      assert.deepStrictEqual(result, true);
+    });
+    it('returns false when user has write permission to repo and attempts to delete a repo', async () => {
+      const result = await repo.allowedToPerformTicket('delete-repo', 'write');
+      assert.deepStrictEqual(result, false);
+    });
+    it('returns true when user has admin permission to repo and attempts to delete a repo', async () => {
+      const result = await repo.allowedToPerformTicket('delete-repo', 'admin');
+      assert.deepStrictEqual(result, true);
+    });
+  });
+
+  describe('performTicketIfAllowed', async () => {
+    let fullName;
+    let testName;
+    let stubbedPermissions;
+    let stubbedTestDeletion;
+
+    beforeEach(() => {
+      stubbedPermissions = sinon.stub(repo, 'allowedToPerformTicket').callsFake((action, permission) => {
+        // permission would normally be write, admin, etc. For this mock just setting to 'permitted' or other for the purpose of testing.
+        if (permission === 'permitted') return true;
+        return false;
+      });
+      stubbedTestDeletion = sinon.stub(deleter, 'deleteTest').callsFake((client, name, test) => {
+        fullName = name;
+        testName = test;
+      });
+    });
+
+    afterEach(() => {
+      stubbedPermissions.restore();
+      stubbedTestDeletion.restore();
+      fullName = null;
```

Comment:
this approach works, but I think you could simplify code simply by using:

```
spy.calledWith(arg1, arg2, ...);
```

basically, sinon has functionality built in similar to what you're doing here.